src/mem/mem_pool.c

```diff
@@ -0,0 +1,251 @@
+ *
+ * This function initializes a memory pool with a specified number of elements,
+ * each of a given size. Optionally, a destructor callback can be provided
+ * to handle cleanup when a member is released or pool is destroyed.
+ *
+ * @param  poolp    Pointer to the memory pool pointer to be initialized.
+ * @param  nmemb    Number of elements to allocate in the pool.
+ * @param  membsize Size of each element in the pool.
+ * @param  dh       Optional destructor callback for pool cleanup (can be
+ *                  NULL).
+ *
+ * @return 0 for success, otherwise error code
+ */
+int mem_pool_alloc(struct mem_pool **poolp, size_t nmemb, size_t membsize,
+		   mem_destroy_h *dh)
+{
+	int err;
+
+	if (!poolp || !nmemb || !membsize)
+		return EINVAL;
+
+	struct mem_pool *p = mem_zalloc(sizeof(struct mem_pool), NULL);
+	if (!p)
+		return ENOMEM;
+
+	p->nmemb    = nmemb;
+	p->membsize = membsize;
+	p->membdh   = dh;
+	p->objs	    = mem_zalloc(nmemb * sizeof(struct mem_pool_entry), NULL);
+	if (!p->objs) {
+		err = ENOMEM;
+		goto error;
+	}
+
+	mem_destructor(p, mem_pool_destroy);
+
+	for (size_t i = 0; i < nmemb; i++) {
+		p->objs[i].member = mem_zalloc(membsize, dh);
+		if (!p->objs[i].member) {
+			err = ENOMEM;
+			goto error;
+		}
+	}
+
+	*poolp = p;
+
+	return 0;
+
+error:
+	mem_deref(p);
+	return err;
+}
+
+
+/**
+ * @brief Extend an existing memory pool.
+ *
+ * Adds additional elements to an existing memory pool.
+ *
+ * @note This function is not thread-safe. Concurrent access to the memory
+ * pool while extending it may result in undefined behavior. Ensure that
+ * proper synchronization mechanisms are in place if this function is used
+ * in a multithreaded context.
+ *
+ * @param pool Pointer to the memory pool to extend.
+ * @param num  Number of additional elements to add to the pool.
+ *
+ * @return 0 for success, otherwise error code
+ */
+int mem_pool_extend(struct mem_pool *pool, size_t num)
+{
+	if (!pool || !num)
+		return EINVAL;
+
+	size_t nmemb = pool->nmemb + num;
+
+	struct mem_pool_entry *objs;
+	objs = mem_zalloc(nmemb * sizeof(struct mem_pool_entry), NULL);
+	if (!objs)
+		return ENOMEM;
+
+	/* Copy old members */
+	size_t i = 0;
+	for (; i < pool->nmemb; i++) {
+		objs[i].member = pool->objs[i].member;
+		re_atomic_rlx_set(&objs[i].used,
+				  re_atomic_rlx(&pool->objs[i].used));
+	}
+
+	/* Allocate new members */
+	for (; i < nmemb; i++) {
+		objs[i].member = mem_zalloc(pool->membsize, NULL);
+		if (!objs[i].member) {
+			mem_deref(objs);
+			return ENOMEM;
+		}
+	}
+
+	mem_deref(pool->objs);
+	pool->objs  = objs;
+	pool->nmemb = nmemb;
+
+	return 0;
+}
+
+
+/**
+ * @brief Borrow an entry from the memory pool.
+ *
+ * Retrieves an unused entry from the memory pool for temporary use.
+ *
+ * @param pool Pointer to the memory pool.
+ *
+ * @return Pointer to a memory pool entry, or NULL if no entries are available.
+ */
+struct mem_pool_entry *mem_pool_borrow(struct mem_pool *pool)
+{
+	if (!pool)
+		return NULL;
+
```

Comment:
- Is it `O(1)`?
- Instead of multiple atomic operations, I think the whole loop should be atomic.